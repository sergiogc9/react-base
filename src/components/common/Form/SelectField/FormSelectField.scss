@import "@src/scss/_globals.scss";
@import "@src/scss/_mixins.scss";
@import "@src/scss/_variables.scss";

.discover-form-select-field {
	@include selectInput("darker");
	label {
		font-size: 1.2rem;
	}

	ul.md-list {
		width: 100%;
		top: 100%;
		li.md-list-item {
			> div {
				padding: 1rem 2rem;
				.md-text {
					font-size: 1.2rem;
				}
			}
		}
	}
}
