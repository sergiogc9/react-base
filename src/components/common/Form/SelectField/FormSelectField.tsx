import React from 'react';
import SelectField from 'react-md/lib/SelectFields/SelectField';
import get from 'lodash/get';
import has from 'lodash/has';

import { withT, TProps } from '@src/lib/i18n';
import { SelectFieldElement } from '@src/types/form';
import { formFieldBaseValidation } from '@src/components/common/Form/Form';

import './FormSelectField.scss';
import { useForceFieldValue } from '@src/lib/hooks';

export type ComponentProps = TProps & {
	element: SelectFieldElement,
	forceValue?: string,
	error: boolean,
	onSelectOption: (elementId: string, value: string, isValid: boolean) => void
};

const __validateSelectField = (element: SelectFieldElement, newValue: string) => {
	if (!formFieldBaseValidation(element, newValue)) return false;
	if (!element.validations) return true;
	return true;
};

const FormSelectField = (props: ComponentProps) => {
	const { t, element, forceValue, error, onSelectOption } = props;

	const [value, setValue] = React.useState(element.defaultValue);
	useForceFieldValue(forceValue, null, element, setValue, onSelectOption, __validateSelectField);

	const __onChangeHandler = React.useCallback((val: string | number) => {
		setValue(val.toString());
		onSelectOption(element.id, val.toString(), __validateSelectField(element, val.toString()));
	}, [element, onSelectOption]);

	const items = React.useMemo(() => element.items.map(item => ({ label: item.label, value: item.value, key: item.value })), [element.items]);

	return <SelectField
		id={`formSelectField-${element.id}`}
		className={`discover-form-select-field ${element.className || ""}`}
		label={has(element, 'label.key') ? t(get(element, 'label.key')) : element.label}
		defaultValue={element.defaultValue}
		value={value}
		menuItems={items}
		onChange={__onChangeHandler}
		required={!!element.required}
		error={error}
		disabled={element.disabled}
		fullWidth={true}
		{...element.inputProps}
	/>;
};

export default withT(React.memo(FormSelectField));
