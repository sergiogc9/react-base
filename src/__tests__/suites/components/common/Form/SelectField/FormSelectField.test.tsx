import React from "react";
import { mount } from "enzyme";
import { act } from 'react-dom/test-utils';

import FormSelectField from "@src/components/common/Form/SelectField/FormSelectField";
import { ComponentProps } from "@src/components/common/Form/SelectField/FormSelectField";
import TestHelper from "@src/__tests__/utils/Helper/TestHelper";

const element = TestHelper.selectFieldElement;

const items = TestHelper.formFieldItems;

let onChangeMock = jest.fn();
describe("Select field common component", () => {
	let wrapper: any;

	const updateComponent = (props: Partial<ComponentProps> = {}) => {
		act(() => {
			wrapper = mount(
				<FormSelectField
					element={element}
					error={false}
					onSelectOption={onChangeMock}
					{...props}
				/>);
		});
	};

	beforeEach(() => {
		onChangeMock = jest.fn();
		updateComponent();
	});

	it("Component default snapshot", () => {
		expect(wrapper.html()).toMatchSnapshot();
	});

	it("Component snapshot with label with key", () => {
		updateComponent({ element: { ...element, label: { key: "lang_key" } } });
		expect(wrapper.html()).toMatchSnapshot();
	});

	it("Component snapshot with items", () => {
		updateComponent({ element: { ...element, items } });
		expect(wrapper.html()).toMatchSnapshot();
	});

	it("Component on blur text input", () => {
		act(() => wrapper.find("SelectField").at(0).props().onChange("newVal"));
		expect(wrapper.html()).toMatchSnapshot();
		expect(onChangeMock).toHaveBeenCalledTimes(1);
	});

	it("Component on validate base validation", () => {
		updateComponent({ element: { ...element, validations: { fn: () => false } } });
		act(() => wrapper.find("SelectField").at(0).props().onChange("newValue"));
		expect(onChangeMock).toHaveBeenCalledWith(element.id, "newValue", false);
	});

	it("Component on validate correct", () => {
		updateComponent({ element: { ...element, validations: {} } });
		act(() => wrapper.find("SelectField").at(0).props().onChange("12345"));
		expect(onChangeMock).toHaveBeenCalledWith(element.id, "12345", true);
	});

	it("Component snapshot after on force update", () => {
		wrapper.setProps({ forceValue: "forced!" });
		expect(onChangeMock).toHaveBeenCalledWith(element.id, "forced!", true);
		expect(wrapper.html()).toMatchSnapshot();
	});
});
