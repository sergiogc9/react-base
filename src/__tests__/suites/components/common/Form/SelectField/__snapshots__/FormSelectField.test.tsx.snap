// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select field common component Component default snapshot 1`] = `"<div id=\\"formSelectField-selectFieldId-menu\\" aria-haspopup=\\"true\\" aria-expanded=\\"false\\" class=\\"md-layover md-layover--simplified md-full-width md-menu-container md-menu--select-field react-base-form-select-field \\"><div class=\\"md-select-field__toggle\\"><label for=\\"formSelectField-selectFieldId\\" class=\\"md-floating-label md-floating-label--inactive md-floating-label--inactive-sized md-text--secondary\\">Text label</label><div id=\\"formSelectField-selectFieldId-toggle\\" role=\\"listbox\\" tabindex=\\"0\\" aria-pressed=\\"false\\" class=\\"md-paper md-paper--0 md-fake-btn md-pointer--hover md-fake-btn--no-outline md-select-field\\"><div class=\\"md-icon-separator md-text-field md-text-field--floating-margin md-select-field--text-field\\"><span class=\\"md-icon-text\\"></span><i class=\\"md-icon material-icons\\">arrow_drop_down</i></div><hr class=\\"md-divider md-divider--text-field md-divider--expand-from-left md-divider--select-field\\"><input type=\\"hidden\\" id=\\"formSelectField-selectFieldId\\" value=\\"\\"></div></div></div>"`;

exports[`Select field common component Component on blur text input 1`] = `"<div id=\\"formSelectField-selectFieldId-menu\\" aria-haspopup=\\"true\\" aria-expanded=\\"false\\" class=\\"md-layover md-layover--simplified md-full-width md-menu-container md-menu--select-field react-base-form-select-field \\"><div class=\\"md-select-field__toggle\\"><label for=\\"formSelectField-selectFieldId\\" class=\\"md-floating-label md-floating-label--inactive md-floating-label--inactive-sized md-text--secondary\\">Text label</label><div id=\\"formSelectField-selectFieldId-toggle\\" role=\\"listbox\\" tabindex=\\"0\\" aria-pressed=\\"false\\" class=\\"md-paper md-paper--0 md-fake-btn md-pointer--hover md-fake-btn--no-outline md-select-field\\"><div class=\\"md-icon-separator md-text-field md-text-field--floating-margin md-select-field--text-field\\"><span class=\\"md-icon-text md-drop-enter\\"></span><i class=\\"md-icon material-icons\\">arrow_drop_down</i></div><hr class=\\"md-divider md-divider--text-field md-divider--expand-from-left md-divider--select-field\\"><input type=\\"hidden\\" id=\\"formSelectField-selectFieldId\\" value=\\"newVal\\"></div></div></div>"`;

exports[`Select field common component Component snapshot after on force update 1`] = `"<div id=\\"formSelectField-selectFieldId-menu\\" aria-haspopup=\\"true\\" aria-expanded=\\"false\\" class=\\"md-layover md-layover--simplified md-full-width md-menu-container md-menu--select-field react-base-form-select-field \\"><div class=\\"md-select-field__toggle\\"><label for=\\"formSelectField-selectFieldId\\" class=\\"md-floating-label md-floating-label--inactive md-floating-label--inactive-sized md-text--secondary\\">Text label</label><div id=\\"formSelectField-selectFieldId-toggle\\" role=\\"listbox\\" tabindex=\\"0\\" aria-pressed=\\"false\\" class=\\"md-paper md-paper--0 md-fake-btn md-pointer--hover md-fake-btn--no-outline md-select-field\\"><div class=\\"md-icon-separator md-text-field md-text-field--floating-margin md-select-field--text-field\\"><span class=\\"md-icon-text md-drop-enter\\"></span><i class=\\"md-icon material-icons\\">arrow_drop_down</i></div><hr class=\\"md-divider md-divider--text-field md-divider--expand-from-left md-divider--select-field\\"><input type=\\"hidden\\" id=\\"formSelectField-selectFieldId\\" value=\\"forced!\\"></div></div></div>"`;

exports[`Select field common component Component snapshot with items 1`] = `"<div id=\\"formSelectField-selectFieldId-menu\\" aria-haspopup=\\"true\\" aria-expanded=\\"false\\" class=\\"md-layover md-layover--simplified md-full-width md-menu-container md-menu--select-field react-base-form-select-field \\"><div class=\\"md-select-field__toggle\\"><label for=\\"formSelectField-selectFieldId\\" class=\\"md-floating-label md-floating-label--inactive md-floating-label--inactive-sized md-text--secondary\\">Text label</label><div id=\\"formSelectField-selectFieldId-toggle\\" role=\\"listbox\\" tabindex=\\"0\\" aria-pressed=\\"false\\" class=\\"md-paper md-paper--0 md-fake-btn md-pointer--hover md-fake-btn--no-outline md-select-field\\"><div class=\\"md-icon-separator md-text-field md-text-field--floating-margin md-select-field--text-field\\"><span class=\\"md-icon-text\\"></span><i class=\\"md-icon material-icons\\">arrow_drop_down</i></div><hr class=\\"md-divider md-divider--text-field md-divider--expand-from-left md-divider--select-field\\"><input type=\\"hidden\\" id=\\"formSelectField-selectFieldId\\" value=\\"\\"></div></div></div>"`;

exports[`Select field common component Component snapshot with label with key 1`] = `"<div id=\\"formSelectField-selectFieldId-menu\\" aria-haspopup=\\"true\\" aria-expanded=\\"false\\" class=\\"md-layover md-layover--simplified md-full-width md-menu-container md-menu--select-field react-base-form-select-field \\"><div class=\\"md-select-field__toggle\\"><label for=\\"formSelectField-selectFieldId\\" class=\\"md-floating-label md-floating-label--inactive md-floating-label--inactive-sized md-text--secondary\\">lang_key</label><div id=\\"formSelectField-selectFieldId-toggle\\" role=\\"listbox\\" tabindex=\\"0\\" aria-pressed=\\"false\\" class=\\"md-paper md-paper--0 md-fake-btn md-pointer--hover md-fake-btn--no-outline md-select-field\\"><div class=\\"md-icon-separator md-text-field md-text-field--floating-margin md-select-field--text-field\\"><span class=\\"md-icon-text\\"></span><i class=\\"md-icon material-icons\\">arrow_drop_down</i></div><hr class=\\"md-divider md-divider--text-field md-divider--expand-from-left md-divider--select-field\\"><input type=\\"hidden\\" id=\\"formSelectField-selectFieldId\\" value=\\"\\"></div></div></div>"`;
